File: sdk/apple/Sources/SpryteSDK/UI/SpriteAvatarView.swift
#if canImport(SwiftUI)
import SwiftUI

/// SwiftUI view that renders a single avatar from a sprite sheet.
///
/// Loads the full sprite PNG via AsyncImage (benefiting from URLSession caching),
/// then offsets and clips to show only the cell for the given pubkey.
@available(iOS 15.0, macOS 12.0, *)
public struct SpriteAvatarView: View {
    let sheet: SpriteSheet
    let pubkey: String
    let size: CGFloat
    let circular: Bool

    private let consumer = SpryteConsumer()

    public init(sheet: SpriteSheet, pubkey: String, size: CGFloat = 48, circular: Bool = true) {
        self.sheet = sheet
        self.pubkey = pubkey
        self.size = size
        self.circular = circular
    }

    public var body: some View {
        if let info = consumer.renderInfo(sheet: sheet, pubkey: pubkey, displaySize: size) {
            AsyncImage(url: info.spriteURL) { phase in
                switch phase {
                case .success(let image):
                    image
                        .resizable()
                        .frame(width: info.scaledWidth, height: info.scaledHeight)
                        .offset(x: info.offsetX, y: info.offsetY)
                case .failure:
                    Color.gray
                        .frame(width: size, height: size)
                default:
                    Color.clear
                        .frame(width: size, height: size)
                }
            }
            .frame(width: size, height: size)
            .clipped()
            .clipShape(circular ? AnyShape(Circle()) : AnyShape(Rectangle()))
        }
    }
}

/// Type-erased Shape wrapper for conditional clipping.
@available(iOS 15.0, macOS 12.0, *)
private struct AnyShape: Shape {
    private let pathBuilder: (CGRect) -> Path

    init<S: Shape>(_ shape: S) {
        pathBuilder = { rect in shape.path(in: rect) }
    }

    func path(in rect: CGRect) -> Path {
        pathBuilder(rect)
    }
}
#endif

#if canImport(UIKit)
import UIKit

@available(iOS 15.0, *)
extension UIImageView {
    /// Load and display a sprite avatar for the given pubkey.
    ///
    /// Downloads the sprite sheet image (or uses cached version), then crops
    /// and scales to the correct cell.
    ///
    /// - Parameters:
    ///   - sheet: A loaded SpriteSheet
    ///   - pubkey: The hex pubkey to render
    ///   - displaySize: Avatar size in points (default: 48)
    public func setSpriteAvatar(sheet: SpriteSheet, pubkey: String, displaySize: CGFloat = 48) {
        let consumer = SpryteConsumer()
        guard let info = consumer.renderInfo(sheet: sheet, pubkey: pubkey, displaySize: displaySize) else {
            self.image = nil
            return
        }

        // Find the entry to get the original cell coordinates
        guard let entry = sheet.mapping.mapping[pubkey] else { return }

        Task {
            do {
                let (data, _) = try await URLSession.shared.data(from: info.spriteURL)
                guard let fullImage = UIImage(data: data),
                      let cgImage = fullImage.cgImage else { return }

                // Crop to the cell in the original image
                let cropRect = CGRect(
                    x: entry.x,
                    y: entry.y,
                    width: sheet.mapping.cellDimensions.width,
                    height: sheet.mapping.cellDimensions.height
                )
                guard let croppedCG = cgImage.cropping(to: cropRect) else { return }
                let cropped = UIImage(cgImage: croppedCG)

                await MainActor.run {
                    self.image = cropped
                    self.contentMode = .scaleAspectFill
                    self.clipsToBounds = true
                    self.layer.cornerRadius = displaySize / 2
                }
            } catch {
                // Silently fail — the image view remains unchanged
            }
        }
    }
}
#endif

File: sdk/android/src/main/kotlin/io/spryte/sdk/ui/SpriteAvatarView.kt
package io.spryte.sdk.ui

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import io.spryte.sdk.SpryteConsumer
import io.spryte.sdk.SpriteSheet

/**
 * Jetpack Compose component that renders a single avatar from a sprite sheet.
 *
 * Loads the full sprite PNG via Coil (benefiting from its disk/memory cache so the
 * sheet image is shared across all avatar instances), then offsets and clips to show
 * only the cell for the given pubkey.
 *
 * @param sheet A loaded SpriteSheet
 * @param pubkey The hex pubkey to render
 * @param displaySize Avatar size in dp (default: 48)
 * @param modifier Optional Compose modifier
 * @param circular Whether to clip the avatar as a circle (default: true)
 */
@Composable
fun SpriteAvatarView(
    sheet: SpriteSheet,
    pubkey: String,
    displaySize: Int = 48,
    modifier: Modifier = Modifier,
    circular: Boolean = true,
) {
    val consumer = SpryteConsumer()
    val renderInfo = consumer.getRenderInfo(sheet, pubkey, displaySize) ?: return

    val density = LocalDensity.current
    val sizeDp = displaySize.dp
    val offsetXDp = with(density) { renderInfo.offsetX.toDp() }
    val offsetYDp = with(density) { renderInfo.offsetY.toDp() }
    val scaledWidthDp = with(density) { renderInfo.scaledWidth.toDp() }
    val scaledHeightDp = with(density) { renderInfo.scaledHeight.toDp() }

    val clipModifier = if (circular) Modifier.clip(CircleShape) else Modifier

    Box(
        modifier = modifier
            .size(sizeDp)
            .then(clipModifier)
    ) {
        AsyncImage(
            model = sheet.spriteUrl,
            contentDescription = "Avatar for $pubkey",
            contentScale = ContentScale.FillBounds,
            modifier = Modifier
                .size(scaledWidthDp, scaledHeightDp)
                .offset(offsetXDp, offsetYDp)
        )
    }
}
